cmake: dts: move to specifying shield on the command line

galak opened the pull request:
Rather than specifying SHIELD via Kconfig, we move it to being specified
via the command line, similar to board.

Fixes #12102

erwango commented on samples/shields/x_nucleo_iks01a1/prj.conf:
```diff
@@ -1,2 +1 @@
 CONFIG_STDOUT_CONSOLE=y
-CONFIG_SHIELD_X_NUCLEO_IKS01A1=y
```
Besides providing a way to select the matching overlay this config symbol was used to do the shield configuration in Kconfig.defconfig and Kconfig.shield.
So, we have 2 options:
- Benefit in application from provided configuration provided by shield. So keep this line.
- Argue that board + shield = new board. New board should then adhere to default configuration guidelines and should let features activation up to application, so remove Kconfig.defconfig and Kconfig.shield and add equivalent configuration in place of shield flag:
```
CONFIG_I2C=y
CONFIG_I2C_STM32_INTERRUPT=y
CONFIG_SENSOR=y
CONFIG_HTS221=y
CONFIG_LPS25HB=y
CONFIG_LSM6DS0=y
CONFIG_LIS3MDL=y
CONFIG_HTS221_TRIGGER_NONE=y
CONFIG_LIS3MDL_TRIGGER_NONE=y
```
I find the second approach more coherent.

erwango commented on samples/shields/x_nucleo_iks01a1/sample.yaml:
```diff
@@ -5,3 +5,5 @@ tests:
     harness: shield
     tags: shield
     depends_on: arduino_i2c
+    extra_args: SHIELD=x_nucleo_iks01a1
```
No more required if CMakeLists amended

erwango reviewed (dismissed):
This works fine, few comments plus following additions:

In `samples/shields/x_nucleo_iks01a1/CMakeLists.txt`:
```
# This sample is specific to x_nucleo_iks01a1 shield. Enforce -DSHIELD option
set(SHIELD x_nucleo_iks01a1)
```

Doc doc change required in `doc/porting/shields.rst`:
```
Shield activation
*****************

Activate support for a shield by adding the matching -DSHIELD arg to CMake
command

  .. zephyr-app-commands::
     :zephyr-app: your_app
     :gen-args: -DSHIELD=x_nucleo_iks01a1
```

erwango committed:
cmake: dts: move to specifying shield on the command line

Rather than specifying SHIELD via Kconfig, we move it to being
specified via the command line, similar to board.

So we can do:

  -DSHIELD=x_nucleo_iks01a1

or, for multiple shields:

  -DSHIELD="x_nucleo_iks01a1 frdm_kw41z"

Following cmake change, update x_nucleo_iks01a1 sample in order
not to enable CONFIG option anymore but set SHIELD cmake option.

Last, update documentation to reflect this change.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

erwango committed:
boards/shields: Clean up Kconfig'ery

Following change of selection mechanism, clean boards/shields from
Kconfig related code.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

dbkinder reviewed (approved):
doc changes LGTM

SebastianBoe commented:
I need more information to be able to help you.

What seems to be the problem?

galak commented:
> I need more information to be able to help you.
> 
> What seems to be the problem?

I wanted to print ${shield_list} in the usage message, when I try to access it in cmake/usage/usage.cmake its empty.

SebastianBoe commented:
I see.

usage.cmake is executed by it's own CMake process, see https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/usage/CMakeLists.txt#L5.

So it only has access to the variables that has been explicitly exported to it, see https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/usage/CMakeLists.txt#L6 for an example of exporting a variable to usage.cmake.

Also, usage.cmake is executed twice, so any additional variables must be applied both here:

https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/usage/CMakeLists.txt#L6

and here:

https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/extensions.cmake#L1114

galak commented:
> I see.
> 
> usage.cmake is executed by it's own CMake process, see [/cmake/usage/CMakeLists.txt@`master`#L5](https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/usage/CMakeLists.txt#L5).
> 
> So it only has access to the variables that has been explicitly exported to it, see [/cmake/usage/CMakeLists.txt@`master`#L6](https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/usage/CMakeLists.txt#L6) for an example of exporting a variable to usage.cmake.
> 
> Also, usage.cmake is executed twice, so any additional variables must be applied both here:
> 
> [/cmake/usage/CMakeLists.txt@`master`#L6](https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/usage/CMakeLists.txt#L6)
> 
> and here:
> 
> [/cmake/extensions.cmake@`master`#L1114](https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/extensions.cmake#L1114)

Ah thanks, that makes sense now.

galak commented:
@SebastianBoe can you re-review the full PR.  Should be complete now.

SebastianBoe commented on cmake/app/boilerplate.cmake:
```diff
@@ -186,6 +186,66 @@ message(STATUS "Selected BOARD ${BOARD}")
+
+elseif(shield_app_cmake_lists)
+  set(SHIELD ${shield_app_cmake_lists})
+else()
```
drop the else().

A reader might assume some code is missing.

SebastianBoe commented on cmake/app/boilerplate.cmake:
```diff
@@ -210,6 +270,39 @@ foreach(root ${BOARD_ROOT})
+        ${shield_dir}/${shield}/dts_fixup.h
+      )
+    endif()
+    endif()
```
indentation

SebastianBoe commented on cmake/app/boilerplate.cmake:
```diff
@@ -210,6 +270,39 @@ foreach(root ${BOARD_ROOT})
+
+    list(APPEND SHIELD_LIST ${shield})
+    if(DEFINED SHIELD)
+    if(${shield} STREQUAL ${SHIELD})
```
With Kconfig, it was possible to enable multiple shields through Kconfig.

To retain this functionality we need to check if ${shield} is in the list ${SHIELD}.

... and possibly rename from SHIELD to SHIELDS.

galak replied:
How would we specify multiple shields, as a 'space separated' list?

`-DSHIELD="frdm_kw41z x_nucleo_iks01a1"`

SebastianBoe replied:
Yes, CONF_FILE should be able to demonstrate how to do it if you see any problems.

galak replied:
> Yes, CONF_FILE should be able to demonstrate how to do it if you see any problems.

Ok, should now handle this case.

MaureenHelm commented on samples/bluetooth/peripheral_hr/CMakeLists.txt:
```diff
@@ -1,4 +1,14 @@
+if(BOARD STREQUAL mimxrt1020_evk OR
+   BOARD STREQUAL mimxrt1050_evk OR
+   BOARD STREQUAL mimxrt1060_evk OR
+   BOARD STREQUAL frdm_k64f)
```
Well that's unfortunate. :(

SebastianBoe replied:
Eventually we are going to need something more sophisticated than CMake variables to give input to DT.

Like a separate Kconfig root/tree for DT for instance.

MaureenHelm commented on samples/bluetooth/peripheral_hr/sample.yaml:
```diff
@@ -4,10 +4,5 @@ sample:
-    harness: bluetooth
-    extra_args: OVERLAY_CONFIG=overlay-shield_frdm_kw41z.conf
-    platform_whitelist: mimxrt1020_evk mimxrt1050_evk mimxrt1060_evk frdm_k64f
+    platform_whitelist: qemu_cortex_m3 qemu_x86 mimxrt1020_evk mimxrt1050_evk mimxrt1060_evk frdm_k64f
```
Instead of the modification to CMakeLists.txt, could you leave the two tests and set `extra_args: SHIELD=frdm_kw41` in the second one?

galak replied:
I went ahead and merged this as I wanted to close on the base -DSHIELD support this week.  I'll open a new PR for reworking the bluetooth test as you've asked in here.

galak replied:
I opened PR #12585 for the BT Periph Sample change.
